### PR TITLE
hexagonal geometry etc.

### PR DESCRIPTION
**Pull request by mlund**
Before sending a pull request, please ensure that:

- all tests pass, i.e. run `make test`
- all compiler warnings have been fixed
- source files use soft-indendation of four
- no temporary files (.o, .pqr, state etc) are included in the commit

### CODE REVIEW

**Comment by mlund** [comment]
Comments:
- "radius" is not a good keyword for hexagonal/octahedron geometry. Add another internal variable.
- add documentation for new geometries. Here "hexagon" should likely be referred to as "hexagonal prism"
- unittests for octahedron?
- unittests for tabulated potential?
- please check indentation in all files

File: examples/water.yml [hunk 1]
```diff
@@ -18,12 +18,12 @@ energy:
     - isobaric: {P/atm: 1}
     - nonbonded_coulomblj:
         lennardjones: {mixing: LB}
-        coulomb: {type: fanourgakis, epsr: 1, cutoff: 9}
+        coulomb: {type: fanourgakis, epsr: 1, cutoff: 9, debyelength: 1.0} # , ellipse_cutoff: True, ellipse_x: 9, ellipse_y: 9, ellipse_z: 9
```
Comment (mlund) [inline examples/water.yml]: please restore water example; it should represent a simple example of bulk water.

File: src/geometry.cpp [hunk 2]
```diff
@@ -67,10 +69,12 @@ namespace Faunus {
...
-                case SPHERE:   return 4*pc::pi/3*radius*radius*radius;
-                case CUBOID:   return len.x()*len.y()*len.z();
-                case SLIT:     return len.x()*len.y()*len.z() / pbc_disable;
-                case CYLINDER: return pc::pi*radius*radius*len.z();
+                case SPHERE:     return 4*pc::pi/3*radius*radius*radius;
+                case CUBOID:     return len.x()*len.y()*len.z();
+                case SLIT:       return len.x()*len.y()*len.z() / pbc_disable;
+                case CYLINDER:   return pc::pi*radius*radius*len.z();
+		case HEXAGONAL:  return 2.0*std::sqrt(3.0)*radius*radius*len.z();
+		case OCTAHEDRON: return pc::pi*radius*radius*len.z();
```
Comment (mlund) [inline src/geometry.cpp]: indentation

File: src/geometry.h [hunk 3]
```diff
@@ -136,19 +136,49 @@ namespace Faunus {
...
+			a.z() = a.z() - len.z();
+		      if(a.z() < -len_half.z())
+			a.z() = a.z() + len.z();
+		    }
                 } //!< Apply boundary conditions
 
                 inline Point vdist(const Point &a, const Point &b) const override {
                     Point r(a - b);
 
+		    if ( type == CUBOID or type == SPHERE or type == CYLINDER or type == SLIT ) {
```
Comment (mlund) [inline src/geometry.h]: indentation

File: src/geometry.h [hunk 4]
```diff
@@ -267,6 +301,31 @@ namespace Faunus {
                 }
                 CHECK( containerOverlap==false );
             }
+            
+            SUBCASE("hexagonal") {
+                json j = {  {"type","hexagonal"}, {"radius",1.0}, {"length", 1.0/2.0/std::sqrt(3.0)} };
```
Comment (mlund) [inline src/geometry.h]: why "radius"?

File: src/potentials.cpp [hunk 5]
```diff
@@ -122,6 +122,10 @@ void Faunus::Potential::CoulombGalore::from_json(const Faunus::json &j) {
         rc1i = 1/rc;
         epsr = j.at("epsr");
         lB = pc::lB( epsr );
+	ellipse_cutoff = j.value("ellipse_cutoff", false);
+	if(ellipse_cutoff)
+	  Rc_ellipse = Point(j.at("ellipse_x"),j.at("ellipse_y"),j.at("ellipse_z"));
+
```
Comment (mlund) [inline src/potentials.cpp]: is this ready to go into a release?

File: src/potentials.h [hunk 6]
```diff
@@ -413,6 +413,63 @@ namespace Faunus {
...
+
+                    Point force(const Tparticle &a, const Tparticle &b, double r2, const Point &p) {
+                        if (a.id == HTMAid or b.id == HTMAid) {
+                            double r6inv=1/(r2*r2*r2);
+                            return 4*m_charged->operator()(a.id,b.id)*r6inv*p;
+                        }
+                        else return Point(0,0,0);
+                    }
+            };
+
```
Comment (mlund) [inline src/potentials.h]: cleanup after new tabulated potential?

File: src/potentials.h [hunk 7]
```diff
@@ -737,8 +791,16 @@ namespace Faunus {
             class FunctorPotential : public PairPotentialBase {
                 typedef std::function<double(const T&, const T&, const Point&)> uFunc;
                 PairMatrix<uFunc,true> umatrix; // matrix with potential for each atom pair
+                typedef Tabulate::TabulatorBase<double>::data Ttable; // data for tabulated potential
+                Tabulate::Andrea<double> tblt; // tabulated potential
+                struct data {
+                    double rmin2;
+                    double rmax2;
```
Comment (mlund) [inline src/potentials.h]: redundant?

File: src/energy.h [hunk 8]
```diff
@@ -1705,6 +1706,9 @@ namespace Faunus {
                                     if (it.key()=="nonbonded_coulombwca")
                                         push_back<Energy::Nonbonded<Tspace,CoulombWCA>>(it.value(), spc);
 
+                                    if (it.key()=="nonbonded_coulombpolarwca")
+                                        push_back<Energy::Nonbonded<Tspace,CoulombPolarWCA>>(it.value(), spc);
+
```
Comment (mlund) [inline src/energy.h]: cleanup after tab. potential?

File: src/geometry.cpp [hunk 9]
```diff
@@ -53,7 +53,9 @@ namespace Faunus {
                 {"cuboid", CUBOID},
                 {"cylinder", CYLINDER},
                 {"slit", SLIT},
-                {"sphere", SPHERE}
+                {"sphere", SPHERE},
+                {"hexagonal", HEXAGONAL},
```
Comment (mlund) [inline src/geometry.cpp]: "hexprism"?

File: src/geometry.cpp [hunk 10]
```diff
@@ -163,6 +177,24 @@ namespace Faunus {
...
                     break;
+                case OCTAHEDRON:
+                    m.x() = (rand()-0.5) * len.x();
+                    m.y() = (rand()-0.5) * len.y();
+                    m.z() = (rand()-0.5) * len.z() / pbc_disable;
+                    break;
+                case HEXAGONAL:
+		    double Ra = rand();
+		    double Rb = rand();
+		    Point unitvA = Point(sqrt(3.0)/2.0,0.5,0.0);
```
Comment (mlund) [inline src/geometry.cpp]: add `unitvA` etc. as const data members

File: src/geometry.cpp [hunk 11]
```diff
@@ -186,6 +218,17 @@ namespace Faunus {
                     if ( std::fabs(a.y()) > len_half.y()) return true;
                     if ( std::fabs(a.z()) > len_half.z()) return true;
                     break;
+                case HEXAGONAL:
+		    if ( std::fabs(a.z()) > len_half.z()) return true;
+		    if(std::fabs(a.dot(Point(1.0,0.0,0.0))) > len_half.x()) return true;
+		    if(std::fabs(a.dot(Point(0.5,sqrt(3.0)/2.0,0.0))) > len_half.x()) return true;
+		    if(std::fabs(a.dot(Point(-0.5,sqrt(3.0)/2.0,0.0))) > len_half.x()) return true;
+                    break;
```
Comment (mlund) [inline src/geometry.cpp]: indent

File: src/geometry.h [hunk 12]
```diff
@@ -136,19 +136,49 @@ namespace Faunus {
                 void to_json(json &j) const;
 
                 inline void boundary( Point &a ) const override {
-                    if ( std::fabs(a.x()) > len_half.x())
+		    if ( type == CUBOID or type == SPHERE or type == CYLINDER or type == SLIT ) {
```
Comment (mlund) [inline src/geometry.h]: indent

File: src/geometry.h [hunk 13]
```diff
@@ -136,19 +136,49 @@ namespace Faunus {
...
-                    if ( std::fabs(a.z()) > len_half.z())
+		      if ( std::fabs(a.z()) > len_half.z())
                         a.z() -= len.z() * anint(a.z() * len_inv.z());
+		    } else if( type == HEXAGONAL ) {
+		      Point unitvX = Point(1.0,0.0,0.0);
+		      Point unitvY = Point(0.5,sqrt(3.0)/2.0,0.0);
+		      Point unitvZ = Point(-0.5,sqrt(3.0)/2.0,0.0);
+		      if(a.dot(unitvX) > len_half.x())
+			a = a - len.x()*unitvX;
+		      if(a.dot(unitvX) < -len_half.x())
```
Comment (mlund) [inline src/geometry.h]: evaluate `a.dot(unitvX` etc only once and save

File: src/geometry.cpp (hunk 9 above)
Comment (mlund) [reply src/geometry.cpp]: perhaps "hexagonal" is good!

File: src/geometry.h (hunk 13 above)
Comment (bjornstenqvist) [reply src/geometry.h]: Can not do this since 'a' is changing after each evaluated dot-product.

**Pull request merged**